pyanaconda/ui/tui/spokes/installation_progress.py

```diff
@@ -61,74 +59,58 @@ def completed(self):
+            # Instead of updating a progress bar, we just
+            # print a pip but print it without a new line.
+            print('.', flush=True, end='')
 
-        # Grab all messages may have appeared since last time this method ran.
-        while True:
-            # Attempt to get a message out of the queue for how we should update
-            # the progress bar.  If there's no message, don't error out.
-            # Also flush the communication Queue at least once a second and
-            # process it's events so we can react to async evens (like a thread
-            # throwing an exception)
-            while True:
-                try:
-                    (code, args) = q.get(timeout=1)
-                    break
-                except queue.Empty:
-                    pass
-                finally:
-                    loop = App.get_event_loop()
-                    loop.process_signals()
-
-            if code == progressQ.PROGRESS_CODE_INIT:
-                # Text mode doesn't have a finite progress bar
-                pass
-            elif code == progressQ.PROGRESS_CODE_STEP:
-                # Instead of updating a progress bar, we just print a pip
-                # but print it without a new line.
-                print('.', flush=True, end='')
-                # Use _stepped as an indication to if we need a newline before
-                # the next message
-                self._stepped = True
-            elif code == progressQ.PROGRESS_CODE_MESSAGE:
-                # This should already be translated
-                if self._stepped:
-                    # Get a new line in case we've done a step before
-                    self._stepped = False
-                    print('')
-                # Print the progress message.
-                print(args[0], flush=True)
-            elif code == progressQ.PROGRESS_CODE_COMPLETE:
-                # There shouldn't be any more progress updates, so return
-                q.task_done()
-
-                if self._stepped:
-                    print('')
-                return True
-            elif code == progressQ.PROGRESS_CODE_QUIT:
-                sys.exit(args[0])
-
-            q.task_done()
-        return True
+            # Use _stepped as an indication to if we need
+            # a newline before the next message.
+            self._stepped = True
 
     def show_all(self):
         super().show_all()
         from pyanaconda.installation import RunInstallationTask
 
+        # Start the installation task.
         task = RunInstallationTask(
             payload=self.payload,
             ksdata=self.data
         )
+        task.progress_changed_signal.connect(
+            self._on_progress_changed
+        )
         task.start()
 
-        # This will run until we're all done with the install thread.
-        self._update_progress()
+        # This will run until the task is finished.
+        while task.is_running:
+            sleep(1)
```

Comment:
@jkonecny12, is it possible implement a solution without the sleep? Does `simpleline` support something like that?

Reply:
You need to wait on something so if you have a source you can use signals if I'm not mistaken. Do we have something like that?

Reply:
There is `return_after` parameter for this method which will end when a specific signal arrive so if you have that it should be pretty straightforward.

Reply:
Thanks! I have used it.